### PR TITLE
feat: format autocmd events

### PR DESCRIPTION
**Pull request by xiaoshihou514**
#62

### CODE REVIEW

**Comment by xiaoshihou514** [comment]
@glepnir 
- Is passing big tables as autocmd data is a bit too costly? :thinking: 
- Should autocmd have the ability to change formatted text and parsed diagnostics?

It's a really early draft and no tests have been written so bear with me :sweat_smile:

**Comment by xiaoshihou514** [comment]
[Screencast from 2023-09-23 19-19-09.webm](https://github.com/nvimdev/guard.nvim/assets/108414369/e28f222d-f237-448d-aec7-bef29099f551)

Made a little statusline component levereging this feature, sth like this:

```lua
        if data then
            if data.status == "pending" then
                is_formatting = true
            elseif data.status == "done" then
                is_formatting = false
            end
        end
        symbol = is_formatting and " " or " "
```

**Comment by xiaoshihou514** [comment]
Currently there's only `GuardFmt` events, pre-format (before spawning ) sends data with `status="pending"` and `using=<formatter config table>`, post-format (after `update_buffer`) sends data with `status="done"` and `results=new_lines`. So:
- Is sending `new_lines` a bit too expensive?
- Is it necessary to add corresponding `GuardLint` events? Haven't thought of a good use case as of yet

**Comment by xiaoshihou514** [comment]
Lint updates on `GuardFmt` event

[Screencast from 2023-09-23 20-09-01.webm](https://github.com/nvimdev/guard.nvim/assets/108414369/14934d2a-4b33-439b-9f4e-e90bed97f13d)

Current behaviour (note how diagnostics have wrong offsets)

[Screencast from 2023-09-23 20-10-10.webm](https://github.com/nvimdev/guard.nvim/assets/108414369/9775510a-6f8e-495e-8a35-b30a85f55b62)

**Comment by xiaoshihou514** [comment]
I also made `do_lint` update on `GuardFmt` (status=done), but doesn't seem to make much of a difference when I tested it on a really large python file.

**Comment by barrett-ruth** [comment]
Would be nice if you add an example config with the status bar in the docs to demonstrate these features.

**Comment by xiaoshihou514** [comment]
> Would be nice if you add an example config with the status bar in the docs to demonstrate these features.

@barrett-ruth 

```lua
-- Guard setup
vim.opt.rtp:append("/tmp/guard.nvim")
local filetype = require("guard.filetype")
filetype("lua"):fmt({
    cmd = "stylua",
    args = { "-" },
    stdin = true,
})
require("guard").setup()

local is_formatting = false
_G.guard_status = function()
    -- Only display icon if auto-format is enabled
    local au = vim.api.nvim_get_autocmds({
        group = "Guard",
        buffer = 0,
    })
    if filetype[vim.bo.ft] and #au ~= 0 then
        -- show formatting status
        return is_formatting and "" or ""
    end
    return ""
end

-- super simple statusline that uses the above function
vim.cmd("au BufEnter * lua vim.opt.stl = [[%f %h%m ]] .. guard_status()")
-- update statusline on GuardFmt event
vim.api.nvim_create_autocmd("User", {
    pattern = "GuardFmt",
    callback = function(opt)
        is_formatting = opt.data.status == "pending"
        vim.opt.stl = [[%f %h%m ]] .. guard_status()
    end,
})
```
I will put that in ADVANCED.md as an example

File: lua/guard/format.lua [hunk 1]
```diff
@@ -132,13 +132,17 @@ local function do_fmt(buf)
...
-
   local fmt_configs = filetype[vim.bo[buf].filetype].formatter
   local fname = vim.fn.fnameescape(api.nvim_buf_get_name(buf))
   local startpath = vim.fn.expand(fname, ':p:h')
   local root_dir = util.get_lsp_root()
   local cwd = root_dir or uv.cwd()
+  util.doau('GuardFmt', {
+    status = 'pending',
+    using = fmt_configs,
+    range = range and { srow, erow },
```
Comment (glepnir) [inline lua/guard/format.lua:143]: what's the purpose of `range` in data

File: lua/guard/init.lua [hunk 2]
```diff
@@ -42,17 +42,20 @@ local function setup(opt)
 
   local lint = require('guard.lint')
   for ft, conf in pairs(ft_handler) do
+    local lint_events = { 'BufWritePost', 'BufEnter' }
```
Comment (glepnir) [inline lua/guard/init.lua:45]: `BufWritePost` no need and lint should run in GuardFmt Done.

File: lua/guard/format.lua (hunk 1 above)
Comment (xiaoshihou514) [reply lua/guard/format.lua:143]: I'm not sure... Just thought that someone may find it useful.

File: lua/guard/init.lua (hunk 2 above)
Comment (xiaoshihou514) [reply lua/guard/init.lua:45]: I made it run on GuardFmt done if fmt_on_save is set and there is a formatter for this ft, guess I could remove BufWritePost in this case.

**Comment by xiaoshihou514** [comment]
https://github.com/nvimdev/guard.nvim/issues/99#issuecomment-1751732918

Some potential for a `GuardLint` event

File: lua/guard/format.lua [hunk 3]
```diff
@@ -186,12 +189,19 @@ local function do_fmt(buf)
 
     vim.schedule(function()
       if not api.nvim_buf_is_valid(buf) or changedtick ~= api.nvim_buf_get_changedtick(buf) then
+        util.doau('GuardFmt', {
+          status = 'failed',
```
Comment (glepnir) [inline lua/guard/format.lua:193]: could add an field `message = 'buffer changed or invalid`

**Review by glepnir — APPROVED** [review]

**Review by barrett-ruth — APPROVED** [review]
LGTM

**Pull request merged**